feat: Add the otelcol syslog receiver (#2263)
grafana/alloy: docs/sources/reference/components/otelcol/otelcol.receiver.syslog.md

### udp block

The `udp` block configures a UDP syslog server.
The following arguments are supported:

| Name                            | Type     | Description                                                                                                  | Default | Required |
|---------------------------------|----------|--------------------------------------------------------------------------------------------------------------|---------|----------|
| `listen_address`                | `string` | The `<host:port>` address to listen to for syslog messages.                                                  |         | yes      |
| `one_log_per_packet`            | `bool`   | Skip log tokenization, improving performance when messages always contain one log and multiline is not used. | `false` | no       |
| `add_attributes`                | `bool`   | Add net.* attributes to log messages according to OpenTelemetry semantic conventions.                        | `false` | no       |
| `encoding`                      | `string` | The encoding of the syslog messages.                                                                         | `utf-8` | no       |

> | `preserve_leading_whitespaces`  | `bool`   | Preserves leading whitespace in messages when set to `true`                                                  | `false` | no       |
> | `preserve_trailing_whitespaces` | `bool`   | Preserves trailing whitespace in messages when set to `true`                                                 | `false` | no       |

Reviewer: ```suggestion
| `preserve_leading_whitespaces`  | `bool`   | Preserves leading whitespace in messages when set to `true` .                                                 | `false` | no       |
| `preserve_trailing_whitespaces` | `bool`   | Preserves trailing whitespace in messages when set to `true`.                                                 | `false` | no       |
```

Forgot the full stop :-)

|---------------------------------|----------|--------------------------------------------------------------------------------------------------------------|---------|----------|
| `listen_address`                | `string` | The `<host:port>` address to listen to for syslog messages.                                                  |         | yes      |
| `max_log_size`                  | `int`    | The maximum size of a log entry to read before failing.                                                      | `1MiB`  | no       |
| `one_log_per_packet`            | `bool`   | Skip log tokenization, improving performance when messages always contain one log and multiline is not used. | `false` | no       |
| `add_attributes`                | `bool`   | Add net.* attributes to log messages according to OpenTelemetry semantic conventions.                        | `false` | no       |
| `encoding`                      | `string` | The encoding of the syslog messages.                                                                         | `utf-8` | no       |

> | `preserve_leading_whitespaces`  | `bool`   | Preserves leading whitespace in messages when set to `true`                                                  | `false` | no       |
> | `preserve_trailing_whitespaces` | `bool`   | Preserves trailing whitespace in messages when set to `true`                                                 | `false` | no       |

Reviewer: ```suggestion
| `preserve_leading_whitespaces`  | `bool`   | Preserves leading whitespace in messages when set to `true`.                                                  | `false` | no       |
| `preserve_trailing_whitespaces` | `bool`   | Preserves trailing whitespace in messages when set to `true`.                                                 | `false` | no       |
```

---

> canonical: https://grafana.com/docs/alloy/latest/reference/components/otelcol/otelcol.receiver.syslog/

Reviewer 2: @clayton-cornell do we need to add a tag for public preview?

Reviewer: Yes. I missed that  :-(  I saw the shared lookup... forgot to make sure the badge was added.

Parallel to this, we are looking at better ways of adding badges to the docs... so we can avoid HTML spans.

| `protocol`                        | `string` | The syslog protocol that the syslog server supports.               | `rfc5424` | no       |
| `location`                        | `string` | The geographic time zone to use when parsing an RFC3164 timestamp. | `UTC`     | no       |
| `enable_octet_counting`           | `bool`   | Whether to enable RFC6587 octet counting.                          | `false`   | no       |
| `max_octets`                      | `int`    | The maximum octets for messages when octet counting is enabled.    | `8192`    | no       |
| `allow_skip_pri_header`           | `bool`   | Allow parsing records without a priority header.                   | `false`   | no       |
| `non_transparent_framing_trailer` | `string` | The framing trailer when using RFC6587 Non-Transparent-Framing.    | `nil`     | no       |

The `protocol` argument specifies the syslog format supported by the receiver.
`protocol` must be one of `rfc5424`, `rfc3164`

The `location` argument specifies a Time Zone identifier. The available locations depend on the local IANA Time Zone database.
See [this wikipedia entry][tz-wiki] for a non-comprehensive list.

> The `non_transparent_framing_trailer` argument must be one of `LF`, `NUL`.

Reviewer 2: Can we add some context around this? Like how can the default be nil? Even if its a string it should be "" but that seems to not be LF or NUL.

Author: Yeah, that's fair. It's a *string, so it's nil if not set. If it is set, then the character chosen is expected to terminate the message. If not set, it's expected that you're using octet counting or that the syslog frame contains only one message.

Reviewer: What's in the docs makes sense to me. I don't know the specific nuances of a string data type and a value of `nil` though. Whats' there seems to make sense as I read it.

### udp block

The `udp` block configures a UDP syslog server.
The following arguments are supported:

| Name                            | Type     | Description                                                                                                  | Default | Required |
|---------------------------------|----------|--------------------------------------------------------------------------------------------------------------|---------|----------|
| `listen_address`                | `string` | The `<host:port>` address to listen to for syslog messages.                                                  |         | yes      |
| `one_log_per_packet`            | `bool`   | Skip log tokenization, improving performance when messages always contain one log and multiline is not used. | `false` | no       |
| `add_attributes`                | `bool`   | Add net.* attributes to log messages according to OpenTelemetry semantic conventions.                        | `false` | no       |
| `encoding`                      | `string` | The encoding of the syslog messages.                                                                         | `utf-8` | no       |

> | `preserve_leading_whitespaces`  | `bool`   | Preserves leading whitespace in messages when set to `true` .                                                 | `false` | no       |

Reviewer: ```suggestion
| `preserve_leading_whitespaces`  | `bool`   | Preserves leading whitespace in messages when set to `true`.                                                 | `false` | no       |
```

Grrr... those extra spaces are REALLY hard to spot in the browser view.

| `readers`          | `int` | The number of goroutines to concurrently read from the UDP syslog server.        | `1`     | no       |
| `processors`       | `int` | The number of goroutines to concurrently process logs before sending downstream. | `1`     | no       |
| `max_queue_length` | `int` | The maximum number of messages to wait for an available processor.               | `100`   | no       |

If `async` is not set, a single goroutine will read and process messages synchronously.

### tcp block

The `tcp` block configures a TCP syslog server.
The following arguments are supported:

| Name                            | Type     | Description                                                                                                  | Default | Required |
|---------------------------------|----------|--------------------------------------------------------------------------------------------------------------|---------|----------|
| `listen_address`                | `string` | The `<host:port>` address to listen to for syslog messages.                                                  |         | yes      |

> | `max_log_size`                  | `int`    | The maximum size of a log entry to read before failing.                                                      | `1MiB`  | no       |

Reviewer 2: This should be a string.

See the upstream receiver [documentation][encoding-documentation] for more details.

The `max_log_size` argument has a minimum value of `64KiB`

### tls block

The `tls` block configures TLS settings used for a server. If the `tls` block
isn't provided, TLS won't be used for connections to the server.

{{< docs/shared lookup="reference/components/otelcol-tls-server-block.md" source="alloy" version="<ALLOY_VERSION>" >}}

### retry on failure block

The `retry_on_failure` block configures the retry behavior when the receiver encounters an error downstream in the pipeline.
A backoff algorithm is used to delay the retry upon subsequent failures.
The following arguments are supported:

| Name               | Type       | Description                                                                                               | Default      | Required |
|--------------------|------------|-----------------------------------------------------------------------------------------------------------|--------------|----------|
| `enabled`          | `bool`     | If true, the receiver will pause reading a file and attempt to resend the current batch of logs on error. | `false`      | no       |

> | `initial_interval` | `duration` | The time to wait after first failure to retry.                                                            | `1 second`   | no       |

Reviewer 2: Durations should be in  formatted to `1s` or `5m`

---
canonical: https://grafana.com/docs/alloy/latest/reference/components/otelcol/otelcol.receiver.syslog/
description: Learn about otelcol.receiver.syslog
title: otelcol.receiver.syslog
---

# otelcol.receiver.syslog

`otelcol.receiver.syslog` accepts syslog messages over the network and forwards them as logs to other `otelcol.*` components.
It supports syslog protocols [RFC5424][] and [RFC3164][] and can receive data over `TCP` or `UDP`.

> > **NOTE**: `otelcol.receiver.syslog` is a wrapper over the upstream
> > OpenTelemetry Collector `syslog` receiver. Bug reports or feature requests will
> > be redirected to the upstream repository, if necessary.

Reviewer: ```suggestion
{{< admonition type="note" >}}
`otelcol.receiver.syslog` is a wrapper over the upstream OpenTelemetry Collector `syslog` receiver.
Bug reports or feature requests will be redirected to the upstream repository, if necessary.
{{< /admonition >}}
```

> Multiple `otelcol.receiver.syslog` components can be specified by giving them
> different labels.

Reviewer: ```suggestion
You can specify multiple `otelcol.receiver.syslog` components by giving them different labels.
```

[RFC5424]: https://www.rfc-editor.org/rfc/rfc5424
[RFC3164]: https://www.rfc-editor.org/rfc/rfc3164

## Usage

```alloy
otelcol.receiver.syslog "LABEL" {
  tcp { ... }
  udp { ... }

  output {
    logs    = [...]
  }
}
```

> 
> 
> 
> 
> ## Arguments

Reviewer: ```suggestion

## Arguments
```

The following arguments are supported:

| Name                              | Type     | Description                                                        | Default   | Required |
|-----------------------------------|----------|--------------------------------------------------------------------|-----------|----------|
| `protocol`                        | `string` | The syslog protocol that the syslog server supports.               | `rfc5424` | no       |

> | `location`                        | `string` | The geographic time zone to use when parsing an rfc3164 timestamp. | `UTC`     | no       |
> | `enable_octet_counting`           | `bool`   | Whether to enable rfc6587 octet counting.                          | `false`   | no       |
> | `max_octets`                      | `int`    | The maximum octets for messages when octet counting is enabled.    | `8192`    | no       |
> | `allow_skip_pri_header`           | `bool`   | Allow parsing records without a priority header.                   | `false`   | no       |
> | `non_transparent_framing_trailer` | `string` | The framing trailer when using rfc6587 Non-Transparent-Framing.    | `false`   | no       |

Reviewer: ```suggestion
| `location`                        | `string` | The geographic time zone to use when parsing an RFC3164 timestamp. | `UTC`     | no       |
| `enable_octet_counting`           | `bool`   | Whether to enable RFC6587 octet counting.                          | `false`   | no       |
| `max_octets`                      | `int`    | The maximum octets for messages when octet counting is enabled.    | `8192`    | no       |
| `allow_skip_pri_header`           | `bool`   | Allow parsing records without a priority header.                   | `false`   | no       |
| `non_transparent_framing_trailer` | `string` | The framing trailer when using RFC6587 Non-Transparent-Framing.    | `false`   | no       |
```

|---------------------------------|----------|--------------------------------------------------------------------------------------------------------------|---------|----------|
| `listen_address`                | `string` | The `<host:port>` address to listen to for syslog messages.                                                  |         | yes      |
| `one_log_per_packet`            | `bool`   | Skip log tokenization, improving performance when messages always contain one log and multiline is not used. | `false` | no       |
| `add_attributes`                | `bool`   | Add net.* attributes to log messages according to OpenTelemetry semantic conventions.                        | `false` | no       |
| `encoding`                      | `string` | The encoding of the syslog messages.                                                                         | `utf-8` | no       |
| `preserve_leading_whitespaces`  | `bool`   | Trims leading whitespace on messages unless this is set to `true`.                                           | `false` | no       |

> | `preserve_trailing_whitespaces` | `bool`   | Trims trailing whitespace on messages unless this is set to `true`.                                          | `false` | no       |

Reviewer: I wonder... is the description for these two args backwards?

Would something like ``Preserves leading whitespace in messages when set to `true`.`` is a better description?

The `encoding` argument specifies the encoding of the incoming syslog messages.
`encoding` must be one of `utf-8`, `utf-16le`, `utf-16be`, `ascii`, `big5`, `nop`. 

> See the upstream receiver [documentation][encoding-documentation] for more details.

Reviewer: ```suggestion
Refer to the upstream receiver [documentation][encoding-documentation] for more details.
```

| `listen_address`                | `string` | The `<host:port>` address to listen to for syslog messages.                                                  |         | yes      |
| `max_log_size`                  | `int`    | The maximum size of a log entry to read before failing.                                                      | `1MiB`  | no       |
| `one_log_per_packet`            | `bool`   | Skip log tokenization, improving performance when messages always contain one log and multiline is not used. | `false` | no       |
| `add_attributes`                | `bool`   | Add net.* attributes to log messages according to OpenTelemetry semantic conventions.                        | `false` | no       |
| `encoding`                      | `string` | The encoding of the syslog messages.                                                                         | `utf-8` | no       |
| `preserve_leading_whitespaces`  | `bool`   | Trims leading whitespace on messages unless this is set to `true`.                                           | `false` | no       |

> | `preserve_trailing_whitespaces` | `bool`   | Trims trailing whitespace on messages unless this is set to `true`.                                          | `false` | no       |

Reviewer: Same as the previous comment.

### output block

{{< docs/shared lookup="reference/components/output-block.md" source="alloy" version="<ALLOY_VERSION>" >}}

## Exported fields

`otelcol.receiver.syslog` does not export any fields.

## Component health

`otelcol.receiver.syslog` is only reported as unhealthy if given an invalid
configuration.

## Debug information

`otelcol.receiver.syslog` does not expose any component-specific debug
information.

## Debug metrics

`otelcol.receiver.syslog` does not expose any component-specific debug metrics.

## Example

This example proxies syslog messages from the `otelcol.receiver.syslog` receiver to the 
`otelcol.exporter.syslog` component, and then sends them on to a `loki.source.syslog` component
before being logged by a `loki.echo` component. This shows how the `otelcol` syslog components
can be used to proxy syslog messages before sending them to another destination.

Using the `otelcol` syslog components in this way results in the messages being forwarded as sent,
attempting to use the `loki.source.syslog` component for a similar proxy use case requires
careful mapping of any structured data fields through the `otelcol.processor.transform` component. A

> very simple example of that can be seen in the [`otelcol.exporter.syslog`][exporter-examples] documentation.

Reviewer: ```suggestion
very simple example of that can be found in the [`otelcol.exporter.syslog`][exporter-examples] documentation.
```

---
canonical: https://grafana.com/docs/alloy/latest/reference/components/otelcol/otelcol.receiver.syslog/
description: Learn about otelcol.receiver.syslog
title: otelcol.receiver.syslog
---

> # otelcol.receiver.syslog

Reviewer: ```suggestion
<span class="badge docs-labels__stage docs-labels__item">Public preview</span>

# otelcol.receiver.syslog
```

|-----------------------------------|----------|--------------------------------------------------------------------|-----------|----------|
| `protocol`                        | `string` | The syslog protocol that the syslog server supports.               | `rfc5424` | no       |
| `location`                        | `string` | The geographic time zone to use when parsing an RFC3164 timestamp. | `UTC`     | no       |
| `enable_octet_counting`           | `bool`   | Whether to enable RFC6587 octet counting.                          | `false`   | no       |
| `max_octets`                      | `int`    | The maximum octets for messages when octet counting is enabled.    | `8192`    | no       |
| `allow_skip_pri_header`           | `bool`   | Allow parsing records without a priority header.                   | `false`   | no       |
| `non_transparent_framing_trailer` | `string` | The framing trailer when using RFC6587 Non-Transparent-Framing.    | `nil`     | no       |

The `protocol` argument specifies the syslog format supported by the receiver.

> `protocol` must be one of `rfc5424`, `rfc3164`

Reviewer: ```suggestion
`protocol` must be one of `rfc5424` or `rfc3164`
```

The `location` argument specifies a Time Zone identifier. The available locations depend on the local IANA Time Zone database.

> See [this wikipedia entry][tz-wiki] for a non-comprehensive list.

Reviewer: ```suggestion
Refer to the [list of tz database time zones][tz-wiki] in Wikipedia for a non-comprehensive list.
```

We should avoid using `See` for an xref link, and where possible (and where it makes sense) use the article title as the link text.

| `listen_address`                | `string` | The `<host:port>` address to listen to for syslog messages.                                                  |         | yes      |
| `one_log_per_packet`            | `bool`   | Skip log tokenization, improving performance when messages always contain one log and multiline is not used. | `false` | no       |
| `add_attributes`                | `bool`   | Add net.* attributes to log messages according to OpenTelemetry semantic conventions.                        | `false` | no       |
| `encoding`                      | `string` | The encoding of the syslog messages.                                                                         | `utf-8` | no       |
| `preserve_leading_whitespaces`  | `bool`   | Preserves leading whitespace in messages when set to `true`.                                                 | `false` | no       |
| `preserve_trailing_whitespaces` | `bool`   | Preserves trailing whitespace in messages when set to `true`.                                                 | `false` | no       |

The `encoding` argument specifies the encoding of the incoming syslog messages.

> `encoding` must be one of `utf-8`, `utf-16le`, `utf-16be`, `ascii`, `big5`, `nop`. 

Reviewer: ```suggestion
`encoding` must be one of `utf-8`, `utf-16le`, `utf-16be`, `ascii`, `big5`, or `nop`. 
```

The `non_transparent_framing_trailer` argument must be one of `LF`, `NUL`.

> The `non_transparent_framing_trailer` and `enable_octet_counting` arguments cannot be used with a UDP syslog server.

Reviewer: ```suggestion
The `non_transparent_framing_trailer` and `enable_octet_counting` arguments can't be used with a UDP syslog server.
```
